internal/provider/folder_resource.go

```diff
@@ -0,0 +1,268 @@
+	}
+}
+
+func (r *folderResource) Create(
+	ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse,
+) {
+	if r.provider == nil || !r.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var plan Folder
+	diags := req.Config.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var createReq client.CreateFolderRequest
+	createReq.Name = plan.Name.ValueString()
+
+	_, _, err := r.provider.service.CreateFolder(ctx, &createReq)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating folder",
+			fmt.Sprintf("Could not create folder: %s", formatAPIErrorMessage(err)),
+		)
+		return
+	}
+
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (r *folderResource) Read(
+	ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse,
+) {
+	if r.provider == nil || !r.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var state Folder
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if state.ID.IsNull() {
+		return
+	}
+
+	folderID := state.ID.ValueString()
+
+	// In case this was an import, validate the ID format.
+	if !uuidRegex.MatchString(folderID) {
+		resp.Diagnostics.AddError(
+			"Unexpected folder ID format",
+			fmt.Sprintf("'%s' is not a valid folder ID format. Expected UUID.", folderID),
+		)
+		return
+	}
+
+	folderObj, httpResp, err := r.provider.service.GetFolder(ctx, folderID)
+	if err != nil {
+		if httpResp != nil && httpResp.StatusCode == http.StatusNotFound {
+			resp.Diagnostics.AddWarning(
+				"Folder not found",
+				fmt.Sprintf("Folder with ID %s is not found. Removing from state.", folderID))
+			resp.State.RemoveResource(ctx)
+		} else {
+			resp.Diagnostics.AddError(
+				"Error getting folder info",
+				fmt.Sprintf("Unexpected error retrieving folder info: %s", formatAPIErrorMessage(err)))
+		}
+		return
+	}
+
+	loadFolderToTerraformState(folderObj, &state)
+
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (r *folderResource) Update(
+	ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse,
+) {
+	// Get folder specification.
+	var plan Folder
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get current state.
+	var state Folder
+	diags = req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var (
+		folderObj *client.Folder
+		err       error
+	)
+
+	// Rename the folder if new name specified.
+	if !plan.Name.IsNull() && plan.Name != state.Name {
+		apiResp, _, err = r.provider.service.RenameFolder(ctx, &client.RenameFolderRequest{
```

Comment:
This isn't specific to this PR, but I'd much rather see a general-purpose `UpdateFolder` PATCH endpoint than an RPC-style `RenameFolder`. Gives us a lot more flexibility in the future.

Reply:
Thanks for the review! That decision is outside the scope of this PR but I can ensure that the team discusses this on the CCAPI PR. Currently that PR is still a work in progress